### PR TITLE
[fastlane] git: more responsible, Windows-compatible shellout

### PR DESCRIPTION
**Pull request by dotdoom**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

Some git actions (ex.: `changelog_from_git_commits`) were not working on Windows. This have been noticed when our projects got student contributors with Windows workstations. While fixing that, I also fixed shell escaping in `git_helper.rb`.

### Description

* some things are `#shellescape`'d, the others aren't. Make it uniform and let `Actions.sh` handle escaping properly;
* additionally, do not use bquot (`` ` ``). It is not compatible with Windows cmd.exe;
* tested on a real project by running on Windows in AppVeyor with `changelog_from_git_commits`, `number_of_commits` and `last_git_tag`.

### CODE REVIEW

File: fastlane/lib/fastlane/helper/git_helper.rb
Comment (janpio) [inline fastlane/lib/fastlane/helper/git_helper.rb]: This is equivalent to `['git', 'log']`, correct?

File: fastlane/lib/fastlane/helper/git_helper.rb
Comment (janpio) [inline fastlane/lib/fastlane/helper/git_helper.rb]: What does the `*` indicate here? (Yep, my ruby is rudimentary)

File: fastlane/lib/fastlane/helper/git_helper.rb
Comment (janpio) [inline fastlane/lib/fastlane/helper/git_helper.rb]: Are we sure `pretty_format` and `date_format` don't have to be quoted? Always?

File: fastlane/lib/fastlane/helper/git_helper.rb
Comment (janpio) [inline fastlane/lib/fastlane/helper/git_helper.rb]: Should this maybe be extracted to another method for clarity? (Then the command can also be constructed step by step, similar to this current method here.

File: fastlane/lib/fastlane/helper/git_helper.rb
Comment (janpio) [inline fastlane/lib/fastlane/helper/git_helper.rb]: Where did this `shellescape`ing move to?

**Comment by janpio** [comment]
Thanks @dotdoom for taking the initiative and fixing this! I left some (probably pretty dumb) comments with questions, would be neat if you could through them.

You probably noticed that fastlane's own tests "only" check for the generated command, not that the command would work cross platform on Linux and Windows as well :/ This is a problem, but we don't have real end to end tests (yet) unfortunately.
Could you share the `Fastfile` you used to test this via AppVeyor?

File: fastlane/lib/fastlane/helper/git_helper.rb
Comment (dotdoom) [reply fastlane/lib/fastlane/helper/git_helper.rb]: Done. Do you think it needs a separate test?

**Review by dotdoom — COMMENTED** [review]
Here's a PR to another project that starts executing fastlane git actions on Windows:
https://github.com/dasfoo/delern/pull/988/files (look for "if Gem.win_platform?" which is now commented out).

The output of Git commands is used in "flutter build" command, specifically `--build-number` and `--build-name` flags, greppable in [Travis (OSX)](https://travis-ci.org/dasfoo/delern/builds/511600064) and [AppVeyor (Windows)](https://ci.appveyor.com/project/dotdoom/delern/builds/23373876) build logs.

Regarding tests, since `git` commands are rather hermetic, I believe they may be tested on a real fixture repository.  I might be able to lend a hand here, but it may need a major change to `Actions.sh` framework, e.g. adding a special parameter to this method that says that a command is "test-ready", i.e. should be executed even in test mode. I won't do this without previously discussing it with fastlane team!

**Comment by dotdoom** [comment]
I believe the latest test failure on Ubuntu is unrelated to my changes:

```plain
W: Failed to fetch http://deb.debian.org/debian/dists/jessie-updates/main/binary-amd64/Packages  404  Not Found
```

**Comment by janpio** [comment]
Correct, that seems to be a networking or CDN problem with the Debian mirrors :/ Right now I hope this will fix itself soon.

File: fastlane/lib/fastlane/helper/git_helper.rb
Comment (janpio) [reply fastlane/lib/fastlane/helper/git_helper.rb]: If you know about it, yeah ;) Ruby sometimes is not the most discoverable language - but I am of course fine with it.

The reason for switching from `git log` to `git, log` is consistency I guess?

File: fastlane/lib/fastlane/helper/git_helper.rb
Comment (janpio) [reply fastlane/lib/fastlane/helper/git_helper.rb]: Ok, so before we used a joined string as a param, not the "vararged" array.

Comment (janpio) [reply fastlane/lib/fastlane/helper/git_helper.rb]: Is this common enough that most ruby devs will know about this? Otherwise maybe a simple comment above that line?

File: fastlane/lib/fastlane/helper/git_helper.rb
Comment (janpio) [reply fastlane/lib/fastlane/helper/git_helper.rb]: Ah, https://github.com/fastlane/fastlane/blob/30be4096f3c9a84c9fee3cf4a73c71a7508313f4/fastlane/lib/fastlane/helper/sh_helper.rb#L121-L129 always felt a bit "magic" to me but now I think I get it: Array as param, String as param, many params (that then can be shelljoined).

File: fastlane/lib/fastlane/helper/git_helper.rb
Comment (janpio) [reply fastlane/lib/fastlane/helper/git_helper.rb]: Never hurts - if you ask 👍

**Comment by janpio** [comment]
> Regarding tests, since `git` commands are rather hermetic, I believe they may be tested on a real fixture repository. I might be able to lend a hand here, but it may need a major change to `Actions.sh` framework, e.g. adding a special parameter to this method that says that a command is "test-ready", i.e. should be executed even in test mode. I won't do this without previously discussing it with fastlane team!

That sounds awesome. The `git` actions might not be the most complex ones we have, but they would provide a great template to learn from and apply to other areas. And if we catch some more Windows/Linux bugs with it, that would be ace.

Best create a super small issue (you can remove any issue template of course) and describe how you would go at it, then other members of the team can look at it and comment as well. But I think we never downvoted a improvement in our tests 🙈

File: fastlane/lib/fastlane/helper/git_helper.rb
Comment (dotdoom) [reply fastlane/lib/fastlane/helper/git_helper.rb]: Great. If you prefer to rather not have these cryptic constucts, I can also go with plain arrays, and teach rubocop to complain if they are ever used.

Consistency indeed is one of the reasons, but in this case `git log` won't work anymore: as we pass each argument separately, `shelljoin` will consider each array element a **single** argument. This is the power of `shelljoin`, but it also means that `git log` will be shell-escaped and the OS will look for a binary called `git log` (with a space inside).

File: fastlane/lib/fastlane/helper/git_helper.rb
Comment (dotdoom) [reply fastlane/lib/fastlane/helper/git_helper.rb]: Exactly.

It's been a while since I've been paid for writing Ruby code, so I can't tell how common it is nowadays. Note that it's also used in `Actions.sh` itself (both argument list and a method call). I added a comment nevertheless, better safe than sorry.

https://github.com/fastlane/fastlane/blob/30be4096f3c9a84c9fee3cf4a73c71a7508313f4/fastlane/lib/fastlane/helper/sh_helper.rb#L11-L13

File: fastlane/lib/fastlane/helper/git_helper.rb
Comment (dotdoom) [reply fastlane/lib/fastlane/helper/git_helper.rb]: Right, I think it's trying to be compatible with [IO#popen](https://www.rubydoc.info/stdlib/core/IO.popen).

**Comment by dotdoom** [comment]
Regarding the test for the new method: I think my suggestion was a little hasteful. Since it's an internal interface to Actions and not an actual, er, "action", it's not possible to test it with a conventional test (`Fastfile.parse`). Besides, that code is already covered by existing tests, and a new test will become a "change watchdog" which merely triggers on harmless changes to the code which do not actually affect the external interface.

While looking for similar existing issues I found #13227, which this PR fixes. Also filed #14479 for `Actions.sh` testing.

Looking further into the test failure: apparently [Debian "jessie" has been removed from the mirrors network](https://lists.debian.org/debian-devel-announce/2019/03/msg00006.html). This will need a fix to the [image](https://github.com/fastlane/fastlane/blob/30be4096f3c9a84c9fee3cf4a73c71a7508313f4/.circleci/config.yml#L69). Filed https://github.com/milch/fastlane-docker/issues/2 for that.

File: fastlane/lib/fastlane/helper/git_helper.rb
Comment (janpio) [reply fastlane/lib/fastlane/helper/git_helper.rb]: I am the worst benchmark here, as I am really dabbling in a language I haven't really mastered. And `%w(` is google-able enough that this should not be a problem for anyone really.

> but in this case git log won't work anymore:

A good point, I didn't see that.

File: fastlane/lib/fastlane/helper/git_helper.rb
Comment (janpio) [reply fastlane/lib/fastlane/helper/git_helper.rb]: Thanks.

`sh_helper.rb` has been written by @jdee whose ruby is pretty advanced sometimes (at least compare to what I produce) ;) But I think we are good here as well now.

**Comment by janpio** [comment]
Thanks for all that.

> Looking further into the test failure: apparently [Debian "jessie" has been removed from the mirrors network](https://lists.debian.org/debian-devel-announce/2019/03/msg00006.html). This will need a fix to the [image](https://github.com/fastlane/fastlane/blob/30be4096f3c9a84c9fee3cf4a73c71a7508313f4/.circleci/config.yml#L69). Filed [milch/fastlane-docker#2](https://github.com/milch/fastlane-docker/issues/2) for that.

❤️ ❤️ ❤️ 

We can merge with this test failing if need be.

**Review by ohayon — APPROVED** [review]
This seems legit to me!

**Comment by dotdoom** [comment]
Thank you for your friendliness and a very fast review turnaround!

**Comment by dotdoom** [comment]
I had to force-push to rebase on the latest `master`.  Is there anything else I can do for this PR?
Also just in case GitHub buried my comment about live test too deep to be discoverable:

> Here's a PR to another project that starts executing fastlane git actions on Windows:
> https://github.com/dasfoo/delern/pull/988/files (look for "if Gem.win_platform?" which is now commented out).
> 
> The output of Git commands is used in "flutter build" command, specifically `--build-number` and `--build-name` flags, greppable in [Travis (OSX)](https://travis-ci.org/dasfoo/delern/builds/511600064) and [AppVeyor (Windows)](https://ci.appveyor.com/project/dotdoom/delern/builds/23373876) build logs.

**Comment by janpio** [comment]
Nope, you are good - now we're just waiting for one of the committers to be bold and sure enough to hit the "Merge" button (After adding an "Approve" review). (Won't be me this time as the ruby syntax is too new for me)

**Comment by dotdoom** [comment]
Sounds good.  No pressure, please take your time.  I wanted to be sure it's not blocked on me.

**Review by joshdholtz — APPROVED** [review]
This looks really good! The rework is 🔥 Thanks for the contribution!

**Pull request merged**
